File: packages/sdk-go/statsfactory.go
// Package statsfactory provides a Go SDK for the statsfactory analytics platform.
//
// It handles batching, background flushing, structured User-Agent headers, and
// session ID generation. Events are queued locally and sent in batches to the
// statsfactory ingestion API.
//
// Basic usage:
//
//	client := statsfactory.New(statsfactory.Config{
//	    ServerURL:     "https://stats.example.com",
//	    AppKey:        "sf_live_xxxx",
//	    ClientName:    "myapp",
//	    ClientVersion: "1.0.0",
//	})
//	defer client.Close()
//
//	client.Track("page_view", statsfactory.Dims{
//	    "page.path": "/home",
//	})
package statsfactory

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"runtime"
	"sync"
	"time"
)

// Version is the SDK version included in the User-Agent header.
const Version = "0.1.0"

// maxBatchSize is the maximum number of events per API request.
const maxBatchSize = 25

// defaultFlushInterval is the default time between background flushes.
const defaultFlushInterval = 30 * time.Second

// Dims is a map of dimension key-value pairs attached to an event.
// Values can be string, int, float64, or bool.
type Dims map[string]any

// Config configures a statsfactory Client.
type Config struct {
	// ServerURL is the base URL of the statsfactory server (e.g. "https://stats.example.com").
	// Required.
	ServerURL string

	// AppKey is the API key for authentication (sent as Bearer token).
	// Required.
	AppKey string

	// ClientName is the name of the application using this SDK (e.g. "tinct").
	// Used in the User-Agent header.
	ClientName string

	// ClientVersion is the version of the application (e.g. "0.1.27").
	// Used in the User-Agent header.
	ClientVersion string

	// FlushInterval is the time between automatic background flushes.
	// Defaults to 30 seconds.
	FlushInterval time.Duration

	// HTTPClient is an optional custom HTTP client. If nil, http.DefaultClient is used.
	HTTPClient *http.Client

	// SessionID is an optional pre-set session ID. If empty, one is generated
	// automatically using crypto/rand.
	SessionID string

	// OnError is called when a flush fails. If nil, errors are silently discarded.
	OnError func(err error)
}

// event is a single event queued for sending.
type event struct {
	Event      string         `json:"event"`
	Timestamp  string         `json:"timestamp,omitempty"`
	SessionID  string         `json:"session_id,omitempty"`
	DistinctID string         `json:"distinct_id,omitempty"`
	Dimensions map[string]any `json:"dimensions,omitempty"`
}

// ingestRequest is the JSON body sent to POST /v1/events.
type ingestRequest struct {
	Events []event `json:"events"`
}

// ingestResponse is the JSON body returned from POST /v1/events.
type ingestResponse struct {
	Accepted int `json:"accepted"`
	Errors   []struct {
		Index   int    `json:"index"`
		Message string `json:"message"`
	} `json:"errors"`
}

// Client is a statsfactory analytics client that batches and flushes events.
type Client struct {
	cfg       Config
	userAgent string
	sessionID string

	mu     sync.Mutex
	queue  []event
	closed bool

	httpClient *http.Client

	// Background worker
	stopCh chan struct{}
	doneCh chan struct{}
}

// New creates a new statsfactory Client and starts the background flush worker.
func New(cfg Config) *Client {
	if cfg.FlushInterval <= 0 {
		cfg.FlushInterval = defaultFlushInterval
	}

	httpClient := cfg.HTTPClient
	if httpClient == nil {
		httpClient = http.DefaultClient
	}

	sessionID := cfg.SessionID
	if sessionID == "" {
		sessionID = generateSessionID()
	}

	c := &Client{
		cfg:        cfg,
		userAgent:  buildUserAgent(cfg),
		sessionID:  sessionID,
		queue:      make([]event, 0, maxBatchSize),
		httpClient: httpClient,
		stopCh:     make(chan struct{}),
		doneCh:     make(chan struct{}),
	}

	go c.backgroundWorker()

	return c
}

// Track enqueues an event with the given name and dimensions.
// Track is safe for concurrent use. It does not block on network I/O.
func (c *Client) Track(eventName string, dims Dims) {
	c.TrackWithOptions(eventName, dims, TrackOptions{})
}

// TrackOptions provides optional per-event overrides.
type TrackOptions struct {
	// Timestamp overrides the event timestamp. If zero, the server assigns one.
	Timestamp time.Time

	// SessionID overrides the client-level session ID for this event.
	SessionID string

	// DistinctID is an optional identity for the event.
	DistinctID string
}

// TrackWithOptions enqueues an event with additional options.
func (c *Client) TrackWithOptions(eventName string, dims Dims, opts TrackOptions) {
	c.mu.Lock()
	defer c.mu.Unlock()

	if c.closed {
		return
	}

	ev := event{
		Event: eventName,
	}

	if !opts.Timestamp.IsZero() {
		ev.Timestamp = opts.Timestamp.UTC().Format(time.RFC3339)
	}

	sid := c.sessionID
	if opts.SessionID != "" {
		sid = opts.SessionID
	}
	ev.SessionID = sid

	if opts.DistinctID != "" {
		ev.DistinctID = opts.DistinctID
	}

	if len(dims) > 0 {
		ev.Dimensions = make(map[string]any, len(dims))
		for k, v := range dims {
			ev.Dimensions[k] = v
		}
	}

	c.queue = append(c.queue, ev)
}

// Flush sends all queued events to the server. It blocks until complete or
// the context is cancelled.
func (c *Client) Flush(ctx context.Context) error {
	c.mu.Lock()
	batch := c.drainLocked()
	c.mu.Unlock()

	return c.sendBatches(ctx, batch)
}

// Close flushes any remaining events and stops the background worker.
// After Close returns, Track calls are silently dropped.
func (c *Client) Close() error {
	c.mu.Lock()
	if c.closed {
		c.mu.Unlock()
		return nil
	}
	c.closed = true
	batch := c.drainLocked()
	c.mu.Unlock()

	// Stop the background worker.
	close(c.stopCh)
	<-c.doneCh

	// Final flush with a generous timeout.
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()
	return c.sendBatches(ctx, batch)
}

// SessionID returns the current session ID.
func (c *Client) SessionID() string {
	return c.sessionID
}

// QueueLen returns the number of events currently queued (useful for testing).
func (c *Client) QueueLen() int {
	c.mu.Lock()
	defer c.mu.Unlock()
	return len(c.queue)
}

// --- internal ---

// drainLocked moves all queued events out and returns them. Must be called with mu held.
func (c *Client) drainLocked() []event {
	if len(c.queue) == 0 {
		return nil
	}
	batch := c.queue
	c.queue = make([]event, 0, maxBatchSize)
	return batch
}

// sendBatches sends events in chunks of maxBatchSize.
func (c *Client) sendBatches(ctx context.Context, events []event) error {
	for len(events) > 0 {
		n := maxBatchSize
		if n > len(events) {
			n = len(events)
		}
		chunk := events[:n]
		events = events[n:]

		if err := c.sendChunk(ctx, chunk); err != nil {
			return err
		}
	}
	return nil
}

// sendChunk sends a single batch of events to the server.
func (c *Client) sendChunk(ctx context.Context, events []event) error {
	body, err := json.Marshal(ingestRequest{Events: events})
	if err != nil {
		return fmt.Errorf("statsfactory: marshal error: %w", err)
	}

	url := c.cfg.ServerURL + "/v1/events"
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, url, bytes.NewReader(body))
	if err != nil {
		return fmt.Errorf("statsfactory: request error: %w", err)
	}

	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+c.cfg.AppKey)
	req.Header.Set("User-Agent", c.userAgent)

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return fmt.Errorf("statsfactory: send error: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode >= 400 {
		var ir ingestResponse
		_ = json.NewDecoder(resp.Body).Decode(&ir)
		msg := fmt.Sprintf("HTTP %d", resp.StatusCode)
		if len(ir.Errors) > 0 {
			msg = fmt.Sprintf("HTTP %d: %s", resp.StatusCode, ir.Errors[0].Message)
		}
		return fmt.Errorf("statsfactory: server error: %s", msg)
	}

	return nil
}

// backgroundWorker periodically flushes the event queue.
func (c *Client) backgroundWorker() {
	defer close(c.doneCh)

	ticker := time.NewTicker(c.cfg.FlushInterval)
	defer ticker.Stop()

	for {
		select {
		case <-ticker.C:
			c.mu.Lock()
			batch := c.drainLocked()
			c.mu.Unlock()

			if len(batch) > 0 {
				ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
				err := c.sendBatches(ctx, batch)
				cancel()

				if err != nil && c.cfg.OnError != nil {
					c.cfg.OnError(err)
				}
			}

		case <-c.stopCh:
			return
		}
	}
}

// buildUserAgent constructs the structured User-Agent header.
// Format: statsfactory-sdk-go/0.1.0 (clientName/clientVersion; os; arch)
func buildUserAgent(cfg Config) string {
	clientPart := ""
	if cfg.ClientName != "" {
		clientPart = cfg.ClientName
		if cfg.ClientVersion != "" {
			clientPart += "/" + cfg.ClientVersion
		}
	}

	if clientPart != "" {
		return fmt.Sprintf("statsfactory-sdk-go/%s (%s; %s; %s)",
			Version, clientPart, runtime.GOOS, runtime.GOARCH)
	}

	return fmt.Sprintf("statsfactory-sdk-go/%s (%s; %s)",
		Version, runtime.GOOS, runtime.GOARCH)
}

File: packages/sdk-py/statsfactory.py
"""
statsfactory Python SDK — anonymous event telemetry.

Zero external dependencies (stdlib only). Works with Python 3.8+.

Basic usage::

    from statsfactory import StatsFactory

    sf = StatsFactory(server_url="https://stats.example.com", app_key="sf_live_xxxx")

    # Count-only event
    sf.track("release_installed", {"version": "1.2.3", "os": "linux"})

    # Metric event (value for SUM/AVG aggregation)
    sf.track("release_downloads", {"repo": "myrepo", "version": "1.2.3"}, value=42)

    sf.flush()   # or sf.close()
"""

from __future__ import annotations

import json
import os
import secrets
import threading
import time
import urllib.request
import urllib.error
from typing import Any, Callable, Dict, List, Optional, Union

__version__ = "0.1.0"

# ── Types ─────────────────────────────────────────────────────────────────────

DimScalar = Union[str, int, float, bool]
Dims = Dict[str, Union[DimScalar, List[DimScalar]]]

# ── ULID generation ───────────────────────────────────────────────────────────

_ULID_CHARS = "0123456789ABCDEFGHJKMNPQRSTVWXYZ"


def _generate_ulid() -> str:
    """Generate a time-sortable 26-character ULID."""
    ms = int(time.time() * 1000)
    time_part = ""
    for _ in range(10):
        time_part = _ULID_CHARS[ms % 32] + time_part
        ms //= 32
    rand_bytes = secrets.token_bytes(10)
    rand_part = "".join(_ULID_CHARS[b % 32] for b in rand_bytes)
    return time_part + rand_part


def _generate_session_id() -> str:
    return secrets.token_hex(16)


# ── SDK ───────────────────────────────────────────────────────────────────────

class StatsFactory:
    """
    Statsfactory analytics client.

    Thread-safe. Events are queued locally and flushed in background batches.
    """

    MAX_BATCH_SIZE = 25
    DEFAULT_FLUSH_INTERVAL = 30.0  # seconds
    FLUSH_TIMEOUT = 10.0  # seconds

    def __init__(
        self,
        server_url: str,
        app_key: str,
        *,
        client_name: str = "",
        client_version: str = "",
        flush_interval: float = DEFAULT_FLUSH_INTERVAL,
        session_id: str = "",
        on_error: Optional[Callable[[Exception], None]] = None,
    ) -> None:
        if not server_url:
            raise ValueError("server_url is required")
        if not app_key:
            raise ValueError("app_key is required")

        self._server_url = server_url.rstrip("/")
        self._app_key = app_key
        self._client_name = client_name
        self._client_version = client_version
        self._on_error = on_error
        self._session_id = session_id or _generate_session_id()
        self._user_agent = self._build_user_agent()

        self._lock = threading.Lock()
        self._queue: List[Dict[str, Any]] = []
        self._closed = False

        self._flush_interval = flush_interval
        self._timer: Optional[threading.Timer] = None
        if flush_interval > 0:
            self._schedule_flush()

    # ── Public API ────────────────────────────────────────────────────────────

    def track(
        self,
        event_name: str,
        dims: Optional[Dims] = None,
        *,
        value: Optional[float] = None,
        timestamp: Optional[str] = None,
        session_id: Optional[str] = None,
        distinct_id: Optional[str] = None,
    ) -> None:
        """
        Enqueue an event. Does not block on I/O.

        :param event_name: Lowercase alphanumeric + underscores, max 64 chars.
        :param dims: Key-value dimensions.
        :param value: Optional numeric value for SUM/AVG/MIN/MAX aggregation.
                      Absence means count-only (default behaviour unchanged).
        :param timestamp: ISO 8601 timestamp string. Defaults to server time.
        :param session_id: Override the client-level session ID.
        :param distinct_id: User/install identity.
        """
        with self._lock:
            if self._closed:
                return

            ev: Dict[str, Any] = {
                "event": event_name,
                "event_key": _generate_ulid(),
                "session_id": session_id or self._session_id,
            }
            if timestamp:
                ev["timestamp"] = timestamp
            if distinct_id:
                ev["distinct_id"] = distinct_id
            if value is not None:
                ev["value"] = value
            if dims:
                ev["dimensions"] = dict(dims)

            self._queue.append(ev)

    def flush(self) -> None:
        """Flush all queued events synchronously. Raises on network/server errors."""
        batch = self._drain()
        if batch:
            self._send_batches(batch)

    def close(self) -> None:
        """Flush remaining events and stop the SDK. Subsequent track() calls are dropped."""
        with self._lock:
            if self._closed:
                return
            self._closed = True
            if self._timer is not None:
                self._timer.cancel()
                self._timer = None
            batch = self._drain_locked()

        if batch:
            self._send_batches(batch)

    @property
    def session_id(self) -> str:
        return self._session_id

    def queue_length(self) -> int:
        with self._lock:
            return len(self._queue)

    # ── Internal ──────────────────────────────────────────────────────────────

    def _build_user_agent(self) -> str:
        parts = [f"statsfactory-sdk-py/{__version__}"]
        if self._client_name:
            client = self._client_name
            if self._client_version:
                client += "/" + self._client_version
            parts.append(f"({client})")
        return " ".join(parts)

    def _drain(self) -> List[Dict[str, Any]]:
        with self._lock:
            return self._drain_locked()

    def _drain_locked(self) -> List[Dict[str, Any]]:
        batch = self._queue
        self._queue = []
        return batch

    def _schedule_flush(self) -> None:
        self._timer = threading.Timer(self._flush_interval, self._background_flush)
        self._timer.daemon = True
        self._timer.start()

    def _background_flush(self) -> None:
        batch = self._drain()
        if batch:
            try:
                self._send_batches(batch)
            except Exception as exc:
                if self._on_error:
                    self._on_error(exc)
        with self._lock:
            if not self._closed:
                self._schedule_flush()

    def _send_batches(self, events: List[Dict[str, Any]]) -> None:
        while events:
            chunk = events[: self.MAX_BATCH_SIZE]
            events = events[self.MAX_BATCH_SIZE :]
            self._send_chunk(chunk)

    def _send_chunk(self, events: List[Dict[str, Any]]) -> None:
        url = self._server_url + "/v1/events"
        payload = json.dumps({"events": events}).encode()

        req = urllib.request.Request(
            url,
            data=payload,
            headers={
                "Content-Type": "application/json",
                "Authorization": f"Bearer {self._app_key}",
                "User-Agent": self._user_agent,
            },
            method="POST",
        )
        try:
            with urllib.request.urlopen(req, timeout=self.FLUSH_TIMEOUT) as resp:
                if resp.status >= 400:
                    body = resp.read().decode(errors="replace")
                    raise RuntimeError(f"StatsFactory: HTTP {resp.status}: {body}")
        except urllib.error.HTTPError as exc:
            body = exc.read().decode(errors="replace") if exc.fp else ""
            raise RuntimeError(f"StatsFactory: HTTP {exc.code}: {body}") from exc

    # ── Context manager support ───────────────────────────────────────────────

    def __enter__(self) -> "StatsFactory":
        return self

    def __exit__(self, *_: Any) -> None:
        self.close()
